dwave/preprocessing/src/presolveimpl.hpp

```diff
@@ -29,43 +31,206 @@ template <class Bias, class Index = int, class Assignment = double>
+        for (auto& constraint : model_.constraints()) normalization_check_nan(constraint);
+    }
 
-        // One time techniques ----------------------------------------------------
+    static void normalization_check_nan(const dimod::Expression<bias_type, index_type>& expression) {
+        // We only care about the biases, so let's just cast to the base type for speed
+        const dimod::abc::QuadraticModelBase<bias_type, index_type>& base = expression;
 
-        // *-- spin-to-binary
-        technique_spin_to_binary();
-        // *-- remove offsets
-        technique_remove_offsets();
-        // *-- flip >= constraints
-        technique_flip_constraints();
-        // *-- remove self-loops
-        technique_remove_self_loops();
+        for (auto it = base.cbegin_quadratic(); it != base.cend_quadratic(); ++it) {
+            if (std::isnan(it->bias)) throw InvalidModelError("biases cannot be NAN");
+        }
 
-        // Trivial techniques -----------------------------------------------------
+        for (size_type v = 0; v < base.num_variables(); ++v) {
+            if (std::isnan(base.linear(v))) throw InvalidModelError("biases cannot be NAN");
+        }
+
+        if (std::isnan(base.offset())) throw InvalidModelError("biases cannot be NAN");
+    }
+
+    /// Convert any >= constraints into <=.
+    void normalization_flip_constraints() {
+        for (auto& constraint : model_.constraints()) normalization_flip_constraint(constraint);
+    }
+
+    /// Convert a >= constraint into <=.
+    static void normalization_flip_constraint(constraint_type& constraint) {
+        if (constraint.sense() == dimod::Sense::GE) constraint.scale(-1);
+    }
+
+    void normalization_remove_invalid_markers() {
+        std::vector<index_type> discrete;
+        for (size_type c = 0; c < model_.num_constraints(); ++c) {
+            auto& constraint = model_.constraint_ref(c);
+
+            if (!constraint.marked_discrete()) continue;
+
+            // we can check if it's well formed
+            if (constraint.is_onehot()) {
+                discrete.push_back(c);
+            } else {
+                constraint.mark_discrete(false);  // if it's not one-hot, it's not discrete
+            }
+        }
+        // check if they overlap
+        size_type i = 0;
+        while (i < discrete.size()) {
+            // check if ci overlaps with any other constraints
+            auto& constraint = model_.constraint_ref(discrete[i]);
+
+            bool overlap = false;
+            for (size_type j = i + 1; j < discrete.size(); ++j) {
+                if (model_.constraint_ref(discrete[j]).shares_variables(constraint)) {
+                    // we have overlap!
+                    overlap = true;
+                    constraint.mark_discrete(false);
+                    break;
+                }
+            }
+
+            if (overlap) {
+                discrete.erase(discrete.begin() + i);
+                continue;
```

Comment:
I think this is a clearer flow than
```c++
            if (overlap) {
                discrete.erase(discrete.begin() + i);
                continue;
            } else {
                ++i;
            }
```
because it emphasizes that this is not the default case.  There is a name for this pattern but it's escaping me at the moment.

Reply:
yeah, I sneak edited my code snippet likely while you were typing :laughing: